File: unity/Assets/Scripts/Core/Faction.cs
// <copyright file="Faction.cs" company="board.fun game">
//     Trafalgar - Age of Sail. Procedural board.fun RTS.
// </copyright>

namespace Trafalgar.Core
{
    using UnityEngine;

    /// <summary>
    /// The two opposing sides (plus a neutral state used while a ship is sinking).
    /// </summary>
    public enum Faction
    {
        /// <summary>No allegiance (e.g. a wreck mid-sink).</summary>
        Neutral = 0,

        /// <summary>The British / Royal Navy fleet (Player 1).</summary>
        British = 1,

        /// <summary>The Franco-Spanish combined fleet (Player 2).</summary>
        FrancoSpanish = 2,
    }

    /// <summary>
    /// Who issues orders for a given <see cref="Faction"/>.
    /// </summary>
    public enum ControlMode
    {
        /// <summary>Controlled by a human at the table (touch / mouse).</summary>
        Human = 0,

        /// <summary>Controlled by the built-in fleet AI.</summary>
        AI = 1,
    }

    /// <summary>
    /// Convenience helpers for working with <see cref="Faction"/> values.
    /// </summary>
    public static class FactionExtensions
    {
        /// <summary>Gets the opposing fleet for a given faction.</summary>
        /// <param name="faction">The faction to query.</param>
        /// <returns>The enemy faction, or <see cref="Faction.Neutral"/> for neutral.</returns>
        public static Faction Enemy(this Faction faction)
        {
            switch (faction)
            {
                case Faction.British: return Faction.FrancoSpanish;
                case Faction.FrancoSpanish: return Faction.British;
                default: return Faction.Neutral;
            }
        }

        /// <summary>Gets the hull / banner colour used to render a faction's ships.</summary>
        /// <param name="faction">The faction to query.</param>
        /// <returns>A display colour.</returns>
        public static Color BannerColor(this Faction faction)
        {
            switch (faction)
            {
                case Faction.British: return new Color(0.85f, 0.78f, 0.62f);
                case Faction.FrancoSpanish: return new Color(0.72f, 0.36f, 0.34f);
                default: return Color.gray;
            }
        }

        /// <summary>Gets the accent colour (sails / flags / UI) for a faction.</summary>
        /// <param name="faction">The faction to query.</param>
        /// <returns>An accent colour.</returns>
        public static Color AccentColor(this Faction faction)
        {
            switch (faction)
            {
                case Faction.British: return new Color(0.20f, 0.45f, 0.85f);
                case Faction.FrancoSpanish: return new Color(0.90f, 0.55f, 0.20f);
                default: return Color.white;
            }
        }

        /// <summary>Gets a short human-readable name for the faction.</summary>
        /// <param name="faction">The faction to query.</param>
        /// <returns>A display string.</returns>
        public static string DisplayName(this Faction faction)
        {
            switch (faction)
            {
                case Faction.British: return "British";
                case Faction.FrancoSpanish: return "Franco-Spanish";
                default: return "Neutral";
            }
        }
    }
}

File: web/src/core/faction.ts
// The two opposing sides (plus a neutral state used while a ship is sinking) —
// a port of Unity `Core/Faction.cs`.

export enum Faction {
  Neutral = 0,
  British = 1,
  FrancoSpanish = 2,
}

export enum ControlMode {
  Human = 0,
  AI = 1,
}

export function enemyOf(faction: Faction): Faction {
  switch (faction) {
    case Faction.British:
      return Faction.FrancoSpanish;
    case Faction.FrancoSpanish:
      return Faction.British;
    default:
      return Faction.Neutral;
  }
}

/** Packs a 0..1 RGB triple into a 0xRRGGBB integer for PixiJS. */
function rgb(r: number, g: number, b: number): number {
  const ri = Math.round(r * 255);
  const gi = Math.round(g * 255);
  const bi = Math.round(b * 255);
  return (ri << 16) | (gi << 8) | bi;
}

/** Hull / banner colour used to render a faction's ships. */
export function bannerColor(faction: Faction): number {
  switch (faction) {
    case Faction.British:
      return rgb(0.85, 0.78, 0.62);
    case Faction.FrancoSpanish:
      return rgb(0.72, 0.36, 0.34);
    default:
      return 0x808080;
  }
}

/** Accent colour (sails / flags / UI) for a faction. */
export function accentColor(faction: Faction): number {
  switch (faction) {
    case Faction.British:
      return rgb(0.2, 0.45, 0.85);
    case Faction.FrancoSpanish:
      return rgb(0.9, 0.55, 0.2);
    default:
      return 0xffffff;
  }
}

/** Accent colour as a CSS hex string, for the DOM HUD. */
export function accentCss(faction: Faction): string {
  return "#" + accentColor(faction).toString(16).padStart(6, "0");
}

export function displayName(faction: Faction): string {
  switch (faction) {
    case Faction.British:
      return "British";
    case Faction.FrancoSpanish:
      return "Franco-Spanish";
    default:
      return "Neutral";
  }
}
